File: web/js/slicer.js
/**
 * Grid maths - a direct port of src/slicedeck/slicer.py.
 *
 * Kept in lockstep with the Python so the browser demo and the hardware
 * pipeline frame identically: same cover-fit rule, same rounding, same zoom
 * semantics. tests/test_slicer.py is the specification for both.
 */

/**
 * A viewport may not shrink below this many source pixels per cell. Past 1:1 a
 * key is interpolating rather than resolving new detail, but a couple of levels
 * of that is still useful digital zoom; below a handful of pixels there is no
 * information left at all, so the stack stops.
 */
export const MIN_CELL_PX = 8;

export const DECKS = {
  mini: { name: 'Stream Deck Mini', cols: 3, rows: 2, keyPx: 80 },
  mk2: { name: 'Stream Deck MK.2', cols: 5, rows: 3, keyPx: 72 },
  xl: { name: 'Stream Deck XL', cols: 8, rows: 4, keyPx: 96 },
  plus: { name: 'Stream Deck +', cols: 4, rows: 2, keyPx: 120 },
};

export class Box {
  constructor(left, top, right, bottom) {
    this.left = left;
    this.top = top;
    this.right = right;
    this.bottom = bottom;
  }
  get width() { return this.right - this.left; }
  get height() { return this.bottom - this.top; }
}

/**
 * The region of the frame that maps onto the whole deck.
 *
 * Dividing width by columns and height by rows - the obvious approach - gives
 * non-square cells, which a square key then stretches. Cropping to the deck's
 * aspect ratio first keeps every tile square and the mosaic undistorted.
 */
export function sourceRect(frameW, frameH, deck, zoom = null) {
  const left = zoom ? zoom.left : 0;
  const top = zoom ? zoom.top : 0;
  const width = (zoom ? zoom.right : frameW) - left;
  const height = (zoom ? zoom.bottom : frameH) - top;

  const target = deck.cols / deck.rows;
  const current = width / height;

  if (current > target) {
    const newW = Math.round(height * target);
    const inset = Math.floor((width - newW) / 2);
    return new Box(left + inset, top, left + inset + newW, top + height);
  }
  const newH = Math.round(width / target);
  const inset = Math.floor((height - newH) / 2);
  return new Box(left, top + inset, left + width, top + inset + newH);
}

/** Cell rectangles in reading order. Rounding is distributed so cells tile exactly. */
export function cells(frameW, frameH, deck, zoom = null) {
  const rect = sourceRect(frameW, frameH, deck, zoom);
  const out = [];
  for (let row = 0; row < deck.rows; row++) {
    const top = rect.top + Math.round((row * rect.height) / deck.rows);
    const bottom = rect.top + Math.round(((row + 1) * rect.height) / deck.rows);
    for (let col = 0; col < deck.cols; col++) {
      const left = rect.left + Math.round((col * rect.width) / deck.cols);
      const right = rect.left + Math.round(((col + 1) * rect.width) / deck.cols);
      out.push({ row, col, key: row * deck.cols + col, box: new Box(left, top, right, bottom) });
    }
  }
  return out;
}

/**
 * Zoom state: a stack of source rectangles.
 *
 * Pressing a key pushes that key's rectangle, so the entire deck then shows
 * what one key was showing. Because the rectangle lives in source pixels,
 * drilling in reveals real detail instead of upscaling a 96px tile.
 */
export class ZoomStack {
  constructor() { this.stack = []; }

  get current() { return this.stack.length ? this.stack[this.stack.length - 1] : null; }
  get depth() { return this.stack.length; }

  /** How much the current viewport magnifies the full frame. */
  magnification(frameW) {
    return this.current ? frameW / this.current.width : 1;
  }

  /** True once a key shows fewer source pixels than it has key pixels. */
  isInterpolated(frameW, frameH, deck) {
    const grid = cells(frameW, frameH, deck, this.current);
    return grid.length > 0 && grid[0].box.width < deck.keyPx;
  }

  /**
   * Drill into one key. Returns false when the viewport holds too little
   * information to divide again.
   */
  push(frameW, frameH, deck, row, col) {
    const cell = cells(frameW, frameH, deck, this.current)
      .find((c) => c.row === row && c.col === col);
    if (!cell) return false;
    // Look one step ahead: the press only helps if the resulting viewport can
    // still be split into cells that carry pixels.
    const next = cells(cell.box.width, cell.box.height, deck);
    if (Math.min(next[0].box.width, next[0].box.height) < MIN_CELL_PX) return false;
    this.stack.push(cell.box);
    return true;
  }

  pop() { return this.stack.pop() ?? null; }
  clear() { this.stack.length = 0; }

  /** Drop a viewport that no longer fits, e.g. after switching to a smaller source. */
  clampTo(frameW, frameH) {
    const box = this.current;
    if (box && (box.right > frameW || box.bottom > frameH)) {
      this.clear();
      return true;
    }
    return false;
  }
}

File: src/slicedeck/slicer.py
"""Grid math for splitting one frame across the keys of a Stream Deck.

The naive approach - divide width by columns and height by rows - stretches the
image, because a deck's key is square but ``frame_width / cols`` rarely matches
``frame_height / rows``. Everything here works on a *source rectangle* chosen so
each cell is square in source pixels, which keeps the mosaic undistorted.
"""

from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass
from typing import Literal

from PIL import Image

from .config import DeckSpec

Fit = Literal["cover", "contain", "stretch"]

#: A viewport may not shrink below this many source pixels per cell. Past 1:1 a
#: key interpolates rather than resolving new detail, which is still useful
#: digital zoom; below a handful of pixels there is no information left at all.
MIN_CELL_PX = 8


@dataclass(frozen=True)
class Box:
    """Pixel rectangle, left/top inclusive and right/bottom exclusive."""

    left: int
    top: int
    right: int
    bottom: int

    @property
    def width(self) -> int:
        return self.right - self.left

    @property
    def height(self) -> int:
        return self.bottom - self.top

    def as_tuple(self) -> tuple[int, int, int, int]:
        return (self.left, self.top, self.right, self.bottom)


@dataclass(frozen=True)
class Cell:
    row: int
    col: int
    box: Box


def source_rect(
    frame_size: tuple[int, int],
    deck: DeckSpec,
    fit: Fit = "cover",
    zoom: Box | None = None,
) -> Box:
    """Pick the region of the frame that maps onto the whole deck.

    ``cover``   - crop the frame to the deck's aspect ratio (fills every key,
                  loses the edges of the frame).
    ``contain`` - use the whole frame; the caller letterboxes when rendering.
    ``stretch`` - use the whole frame and accept the distortion.

    ``zoom`` restricts the operation to a sub-rectangle first, which is what
    drilling into a single key does.
    """
    frame_w, frame_h = frame_size
    if frame_w <= 0 or frame_h <= 0:
        raise ValueError(f"invalid frame size {frame_size!r}")

    left, top, right, bottom = (
        zoom.as_tuple() if zoom else (0, 0, frame_w, frame_h)
    )
    width, height = right - left, bottom - top
    if width <= 0 or height <= 0:
        raise ValueError("zoom rectangle is empty")

    if fit in ("stretch", "contain"):
        return Box(left, top, right, bottom)

    target = deck.cols / deck.rows
    current = width / height
    if current > target:
        # Too wide: trim the sides.
        new_w = round(height * target)
        inset = (width - new_w) // 2
        return Box(left + inset, top, left + inset + new_w, bottom)
    # Too tall: trim top and bottom.
    new_h = round(width / target)
    inset = (height - new_h) // 2
    return Box(left, top + inset, right, top + inset + new_h)


class Grid:
    """Maps deck keys onto rectangles of a frame."""

    def __init__(self, deck: DeckSpec, fit: Fit = "cover") -> None:
        self.deck = deck
        self.fit = fit

    def index_of(self, row: int, col: int) -> int:
        return row * self.deck.cols + col

    def cells(
        self, frame_size: tuple[int, int], zoom: Box | None = None
    ) -> list[Cell]:
        """Cell rectangles in reading order (row 0 left to right, then row 1)."""
        rect = source_rect(frame_size, self.deck, self.fit, zoom)
        cells: list[Cell] = []
        for row in range(self.deck.rows):
            # Distribute rounding error across cells instead of dropping pixels
            # off the right and bottom edges.
            top = rect.top + round(row * rect.height / self.deck.rows)
            bottom = rect.top + round((row + 1) * rect.height / self.deck.rows)
            for col in range(self.deck.cols):
                left = rect.left + round(col * rect.width / self.deck.cols)
                right = rect.left + round((col + 1) * rect.width / self.deck.cols)
                cells.append(Cell(row, col, Box(left, top, right, bottom)))
        return cells

    def slice(
        self, frame: Image.Image, zoom: Box | None = None
    ) -> Iterator[tuple[Cell, Image.Image]]:
        """Yield each key's cropped and key-sized tile."""
        size = (self.deck.key_px, self.deck.key_px)
        for cell in self.cells(frame.size, zoom):
            tile = frame.crop(cell.box.as_tuple())
            if tile.size != size:
                tile = tile.resize(size, Image.Resampling.LANCZOS)
            yield cell, tile

    def zoom_into(
        self, frame_size: tuple[int, int], row: int, col: int, zoom: Box | None = None
    ) -> Box:
        """The source rectangle for one key - the new viewport after a key press."""
        for cell in self.cells(frame_size, zoom):
            if cell.row == row and cell.col == col:
                return cell.box
        raise IndexError(f"no key at row={row} col={col} on a {self.deck.name}")

    def can_divide(self, box: Box) -> bool:
        """Whether a viewport still holds enough pixels to be split again."""
        cells = self.cells((box.width, box.height))
        return min(cells[0].box.width, cells[0].box.height) >= MIN_CELL_PX

    def is_interpolated(self, frame_size: tuple[int, int], zoom: Box | None = None) -> bool:
        """True once each key is showing fewer source pixels than it has key pixels."""
        return self.cells(frame_size, zoom)[0].box.width < self.deck.key_px
